### PR TITLE
Switch linting to Ruff

### PR DESCRIPTION
**Pull request by akx**
# Description

This PR switches the linting + import formatting for Gradio from flake8 + isort to [Ruff](https://github.com/charliermarsh/ruff).

Ruff is a lot faster than flake8 and isort, and also catches more errors; see some of the fixes done here, and I didn't even configure any [additional lint rules](https://beta.ruff.rs/docs/rules/).

Please see each individual commit and commit message for further detail.

### CODE REVIEW

**Comment by abidlabs** [comment]
Thanks @akx for initiating this PR. Let me discuss with the other maintainers of this repo if we'd like to make this change. Just FYI we prefer if contributors raise major changes as issue so we can discuss there before opening a PR directly.

File: pyproject.toml
Comment (freddyaboulton) [inline pyproject.toml:78]: In my opinion, it's fine if we don't lint the demo files

File: website/check_version.py
Comment (freddyaboulton) [inline website/check_version.py:1]: Also fine if we don't lint the website

**Comment by abidlabs** [comment]
Agree with @freddyaboulton's suggestion about not linting the demos. Because the demos are code that we embed in our documentation and guides, there are times when we don't want to be following pep8 standards, e.g. because it might add too much whitespace. For the website, I don't know, maybe we _should_ actually be linting it -- it might help us catch bugs. WDYT @freddyaboulton @aliabd?

**Comment by akx** [comment]
>  there are times when we don't want to be following pep8 standards, e.g. because it might add too much whitespace. 

`ruff` doesn't do formatting (yet), just linting. If you look at the changes enacted for the demos, they're pretty much just removals of unused imports and variables.

File: scripts/lint_backend.sh
Comment (freddyaboulton) [reply scripts/lint_backend.sh:5]: Ah got it!

File: gradio/__init__.py
Comment (freddyaboulton) [inline gradio/__init__.py:20]: Let's get rid of these duplicate imports?

File: gradio/components.py
Comment (freddyaboulton) [inline gradio/components.py:3076]: Lol nice catch by ruff

**Review by freddyaboulton — COMMENTED** [review]
Thanks @akx ! This looks good to me - just left some minor comments. @abidlabs Any further thoughts before merging?

File: gradio/__init__.py
Comment (akx) [reply gradio/__init__.py:20]: I think `Checkboxgroup` is a legacy compatibility alias for `CheckboxGroup`, so getting rid of it would probably break some code out there.

File: pyproject.toml
Comment (freddyaboulton) [inline pyproject.toml:84]: We can get rid of this since you already added `__all__` right?

File: gradio/__init__.py
Comment (freddyaboulton) [reply gradio/__init__.py:20]: You're right!

File: pyproject.toml
Comment (akx) [reply pyproject.toml:84]: Nope - that was a different module (`client/python/gradio_client/__init__.py`). 

I felt like adding about 80 `__all__` entries would've been a bigger change that can be done separately. 😁

**Review by freddyaboulton — APPROVED** [review]
Thanks for the contribution @akx ! This looks good to me. Will leave open in case @abidlabs wants to make any further comments. Should be good to merge shortly

File: test/test_examples.py
Comment (abidlabs) [inline test/test_examples.py:263]: Was this caught by `ruff` or did you change manually @akx?

**Comment by abidlabs** [comment]
LGTM as well. It looks like `ruff` is catching some additional things that are good to fix. I'm curious if there are things that `ruff` is missing that would be caught by `isort`/`flake`? I don't think anything major and it looks like the other HF libraries are moving towards `ruff` as well, so I'm in favor of making this change

In terms of whether we should `ruff`:
* the demos: I was initially concerned about whitespace but this doesn't seem to have been affected, while other unnecessary imports seem to have removed so it looks like a positive change
* the website: Let's just make sure @aliabd, @aliabid94, and anyone else who works on the website is aware that now they'll have to run `format_backend.sh` to ensure that the website python files are linted

**Comment by abidlabs** [comment]
Merging now, thanks for this nice PR @akx!

**Comment by akx** [comment]
> * they'll have to run `format_backend.sh` to ensure that the website python files are linted

I could follow up with a PR that switches the formatting/linting from those scripts to [`pre-commit`](https://pre-commit.com/), if that helps :)

File: test/test_examples.py
Comment (akx) [reply test/test_examples.py:263]: ruff was complaining about `l` being an ambiguous name, so while renaming, I fixed these up to the equivalent listcomp.

**Pull request merged**
